Fix checkout exception when adding required field to address

fatmaBouchekoua opened the pull request:
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | When you add phone as required fields in Address customers got Exception in checkout when they are already registered with an address without phone number
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/BOOM-2989
| How to test?  | Reinstall PrestaShop then register a customer without adding a phone number to his address, then make it required and try to pass an order with this address.

- [ ] Port on StarterTheme: https://github.com/PrestaShop/StarterTheme/pull/201
<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/7975)
<!-- Reviewable:end -->

fatmaBouchekoua committed:
FO: Fix checkout exception when adding required field

marionf commented:
Hello @fatmaBouchekoua,

![capture du 2017-06-12 14-31-57](https://user-images.githubusercontent.com/13449658/27033984-f5356962-4f7b-11e7-89c5-9c961f234376.png)


Can you please update the warning message for: "Your address is incomplete, please update it." ?

Thank you !

fatmaBouchekoua commented:
Hello @marionf ,
I found this warning message already in use but not shown so I've use it. 
@vincentbz what do you think?

vincentbz commented:
Hi,

I've just seen with Alex and we would prefer to use this new string "Your address is incomplete, please update it.".

Thanks !

fatmaBouchekoua committed:
FO: Add new error message

fatmaBouchekoua commented:
Done.

Thanks!

aleeks commented on classes/AddressChecksumCore.php:
```diff
@@ -49,7 +49,7 @@ public function generateChecksum($address)
             $fields = $address->getFields();
         } catch (Exception $e) {
-            throw new Exception($e->getMessage());
+            throw new Exception($this->trans('Your address is incomplete, please update it.', array(), 'Shop.Notifications.Error'));
```
Hello @fatmaBouchekoua 
Why not : ?
```php
Context::getContext()->getTranslator()->trans('Your address is incomplete, please update it.', array(), 'Shop.Notifications.Error')
```

fatmaBouchekoua replied:
I tried to keep some coherence with other classes. Also, if anyone ever adds a function with translatable strings it will use the trans function directly.

aleeks replied:
In controllers, trans is always accessible but here, we don't have to do that.
You can use $this->getTranslator()->trans(...) if you want but,if we change the "original" trans function, we have to change here to and it's not the good way to do this.
Maybe, the best solution is to do as usual. ping @maximebiloe

maximebiloe replied:
Yes, you shouldn't redefine the `trans()` method. We must have a unique one.
You can keep the `getTranslator()` function but can you change its visibility to `private`, please?

fatmaBouchekoua replied:
Done. Thanks ! 
I've rebased it too.

maximebiloe commented:
Need a little rebase over here

fatmaBouchekoua commented:
Done !

marionf commented:
Hello @fatmaBouchekoua 

The message is well displayed the first time
![pr7975-1](https://user-images.githubusercontent.com/13449658/27434001-47df2e54-5757-11e7-9c0f-c02bef0ac994.png)

But when I select the other address, it no longer appears
![pr7975-2](https://user-images.githubusercontent.com/13449658/27434016-5a666722-5757-11e7-8b89-f580d50e1f54.png)

fatmaBouchekoua commented:
Hello @marionf ,

I've made some changes, can you retest please?

Thanks!

marionf commented:
Hello @fatmaBouchekoua 
The message is still displayed, even after added the phone number in the address.
Also, the button "Continue" should be disabled if the address isn't complete.
And the "edit" button should be in blu
he address isn't complete.
And the "edit" button should be in blue.

![boom-2989](https://user-images.githubusercontent.com/13449658/27831198-e84a58a2-60c9-11e7-9392-155ac6647902.png)

vincentbz commented:
Tested and ok, thanks !

<!--
status: 'QA approved'
-->

marionf commented:
Hello @fatmaBouchekoua, @vincentbz 

It's not ok for me, please see the images below:

![capture du 2017-08-10 11-29-52](https://user-images.githubusercontent.com/13449658/29165912-c06d89f4-7dc4-11e7-9b98-c298369359a6.png)

![capture du 2017-08-10 11-31-22](https://user-images.githubusercontent.com/13449658/29165917-c696785e-7dc4-11e7-8106-052dfd4e6544.png)

![capture du 2017-08-10 11-33-09](https://user-images.githubusercontent.com/13449658/29165926-d0d81caa-7dc4-11e7-8076-1965866a1e69.png)

eternoendless commented on classes/AddressChecksumCore.php:
```diff
@@ -45,12 +46,25 @@ public function generateChecksum($address)
+        try {
+            $fields = $address->getFields();
+        } catch (Exception $e) {
+            throw new Exception($this->getTranslator()->trans('Your address is incomplete, please update it.', array(), 'Shop.Notifications.Error'));
```
When re-throwing exceptions, the previous exception should should always be included in the new one (third parameter on the Exception constructor), so that the original exception can be traced back.

eternoendless replied:
I don't think this is the right place to set up a translated message. You should throw a specific Exception type which can be catched at a higher level (controller or business class), which in turn will display the appropriate message.

eternoendless commented on classes/Customer.php:
```diff
@@ -559,7 +559,7 @@ public static function resetAddressCache($idCustomer, $idAddress)
     public function getAddresses($idLang)
     {
-        $shareOrder = (bool) Context::getContext()->shop->getGroup()->share_order;
+        $shareOrder = (bool)isset(Context::getContext()->shop->getGroup()->share_order) ? Context::getContext()->shop->getGroup()->share_order : false;
```
The `(bool)` is misplaced, it applies to the condition and not to the assignment. It should be placed after the `?`

In addition, you could store `Context::getContext()->shop->getGroup()` in a variable in order to make this line shorter.

eternoendless commented on classes/checkout/CheckoutAddressesStep.php:
```diff
@@ -202,18 +202,37 @@ public function handleRequest(array $requestParams = array())
+        $id_address_invoice = $this->getCheckoutSession()->getIdAddressInvoice();
+        $addressWarning = false;
+        $errorKey = 'address_error';
+        if (isset($this->context->controller->checkoutWarning['address']) &&
```
When writing multiline `if` statements, please keep in mind
* Putting the boolean operators (`&&`, '||`, 'xor') at the **start** of the line
* Optionally, closing the parenthesis in a new line in order to separate the statement from the block:
```php
// hard to read
if ($a === true && 
    $b === false) {
    $c = doSomething();
}

// better
if ($a === true
    && $b === false
) {
    $c = doSomething();
}
```

eternoendless commented on classes/checkout/CheckoutAddressesStep.php:
```diff
@@ -202,18 +202,37 @@ public function handleRequest(array $requestParams = array())
+        $addressWarning = false;
+        $errorKey = 'address_error';
+        if (isset($this->context->controller->checkoutWarning['address']) &&
+            $id_address_delivery == $this->context->controller->checkoutWarning['address']['id_address']) {
```
You should always use strict comparison, and type-cast as needed

eternoendless commented on classes/checkout/CheckoutAddressesStep.php:
```diff
@@ -223,4 +242,29 @@ public function render(array $extraParams = array())
+     * @param $idLang
+     * @return array
+     */
+    private function getChekedCustomerAddresses($idCustomer, $idLang)
```
This method is called `getChekedCustomerAddresses`, but it returns a list of invalid adresses. I think it would be more declarative if it was called `getInvalidCustomerAddresses`

eternoendless commented on classes/AddressChecksumCore.php:
```diff
@@ -45,12 +46,25 @@ public function generateChecksum($address)
+    {
+        return Context::getContext()->getTranslator();
+    }
+
```
This is a bad idea, why should this class need to know about translating?

eternoendless commented on themes/_core/js/checkout-address.js:
```diff
@@ -32,8 +37,66 @@ export default function () {
+  }
+});
+
+function setAddressBox(enabled, id, type) {
```
You should always use 
```js
const setAddressBox = function(enabled, id, type) {
```
and move the function up (see https://github.com/airbnb/javascript#functions--declarations)

eternoendless replied:
This method is missing its JSDoc header

eternoendless commented on themes/_core/js/checkout-address.js:
```diff
@@ -32,8 +37,66 @@ export default function () {
-  $('#delivery-addresses, #invoice-addresses input[type=radio]').on('click', function () {
+  $('#delivery-addresses input[type=radio], #invoice-addresses input[type=radio]').on('click', function () {
+    if ($(".js-address-error" ).length) {
+      var idFailureAddress = $(".js-address-error").prop('id').split('-').pop();
```
You should use `let` instead of `var`

eternoendless commented on themes/_core/js/checkout-address.js:
```diff
@@ -32,8 +37,66 @@ export default function () {
+  $('#id-address-' + type + '-address-' + id + ' a.edit-address').prop('style', 'color: ' + color + ' !important');
+}
+
+function setConfirmAddressesButton() {
```
Missing JSDoc

eternoendless replied:
I don't think the name of this function is declarative regarding what the function actually does. 

"setConfirmAdressesButton" --> Where is it set? Why?

From the looks of it, this function toggles a button between enabled and disable. In that case, I think "toggleConfirmAddresssesButtonDisabled" would be more appropriate.

eternoendless commented on themes/_core/js/checkout-address.js:
```diff
@@ -32,8 +37,66 @@ export default function () {
+  var color = "#7a7a7a";
+  if (enabled) {
+    color = "#2fb5d2";
+  }
```
This block can be merged in a single line

```js
let color = (enabled) ? '#2fb5d2' : '#7a7a7a';
```

eternoendless replied:
Can't this be done with a css class ?

eternoendless commented on themes/_core/js/checkout-address.js:
```diff
@@ -32,8 +37,66 @@ export default function () {
 }
+
+$(window).load(() => {
+  if (useSameAddress == 0) {
```
You should always use strict comparison `===`

eternoendless commented on themes/_core/js/common.js:
```diff
@@ -28,3 +28,23 @@ export function psShowHide () {
+      vars[key] = value !== undefined ? value : '';
+    }
+  );
+  if (paramName) {
```
`paramName !== undefined`

eternoendless commented on classes/checkout/CartChecksum.php:
```diff
@@ -48,10 +48,23 @@ public function generateChecksum($cart)
+                'id_address' => $cart->id_address_delivery,
+                'exception' => $e->getMessage(),
+            );
+        }
```
I don't think it's a good idea to make this method return anything other than a checksum.
You should handle the exception in the Controller instead.

eternoendless commented on themes/_core/js/checkout-address.js:
```diff
@@ -32,8 +37,66 @@ export default function () {
+    $('button[name=confirm-addresses]').prop("disabled", "disabled");
+  } else {
+    $('button[name=confirm-addresses]').prop("disabled", "");
+  }
```
I think `$(".js-address-error:visible").length` should be a value passed as a parameter to the function, so that you could use it in lines 75, 78 and 84:

```js
const toggleConfirmAddresssesButtonDisabled = function(shouldBeDisabled) {
  $('button[name=confirm-addresses]')
    .prop('disabled', shouldBeDisabled ? 'disabled' : '');
}
```

eternoendless commented on themes/_core/js/common.js:
```diff
@@ -28,3 +28,23 @@ export function psShowHide () {
+ * @returns {string|null}
+ */
+export function psGetRequestParameter(paramName) {
+  var vars = {};
```
You should use `let` instead of `var`

eternoendless commented on themes/_core/js/common.js:
```diff
@@ -28,3 +28,23 @@ export function psShowHide () {
+  var vars = {};
+  window.location.href.replace(location.hash, '').replace(
+    /[?&]+([^=&]+)=?([^&]*)?/gi,
+    function (m, key, value) {
```
You should use an arrow function here

eternoendless commented on classes/checkout/CheckoutAddressesStep.php:
```diff
@@ -202,18 +202,37 @@ public function handleRequest(array $requestParams = array())
+            && $id_address_delivery === (int) $this->context->controller->checkoutWarning['address']['id_address']) {
+            $addressWarning = $this->context->controller->checkoutWarning['address'];
+            $errorKey = 'delivery_address_error';
+        } elseif (isset($this->context->controller->checkoutWarning['address'])
```
```php
if (isset($this->context->controller->checkoutWarning['address']) {
  // ...
} elseif (isset($this->context->controller->checkoutWarning['address']) {
  // ...
}
```
Both statements are the same.

eternoendless commented on classes/checkout/CheckoutAddressesStep.php:
```diff
@@ -202,18 +202,37 @@ public function handleRequest(array $requestParams = array())
     public function getTemplateParameters()
     {
+        $id_address_delivery = (int) $this->getCheckoutSession()->getIdAddressDelivery();
+        $id_address_invoice = (int) $this->getCheckoutSession()->getIdAddressInvoice();
```
Please don't use snake_case on new symbols :)

eternoendless commented on controllers/front/OrderController.php:
```diff
@@ -150,7 +151,12 @@ protected function bootstrap()
+        
+        try {
+            $data['checksum'] = $this->cartChecksum->generateChecksum($this->context->cart);
+        } catch (Exception $e) {
```
You shouldn't suppress base exceptions. If you need to, you should throw a specific type of Exception (aka a subclass of Exception) so you catch the exception you're expecting, and not *any exception*.

eternoendless commented on controllers/front/OrderController.php:
```diff
@@ -150,7 +151,12 @@ protected function bootstrap()
-        $data['checksum'] = $this->cartChecksum->generateChecksum($this->context->cart);
+        
+        try {
+            $data['checksum'] = $this->cartChecksum->generateChecksum($this->context->cart);
```
I don't think it should not be up to CartChecksum to verify the integrity of the Cart, because it violates the Single Responsibility Principle.

You should create a `CartValidator` that does this job, with a method `bool isValid($cart)`, and eventually `int[] getInvalidAdresses()` which returns an array of invalid address ids. No need to be throwing exceptions around anymore.

eternoendless commented on themes/_core/js/checkout-address.js:
```diff
@@ -32,8 +36,68 @@ export default function () {
     $('.address-item').removeClass('selected');
     $('.address-item:has(input[type=radio]:checked)').addClass('selected');
+
+    switchConfirmAddresssesButtonState();
```
Before, you were always setting the state to enabled. Now it toggles it alternatively on and off. Are you sure this is what you intended?

mickaelandrieu commented on classes/checkout/CartValidator.php:
```diff
@@ -0,0 +1,110 @@
+
+        return $invalidAddresses;
+    }
+}
```
add an extra line here

fatmaBouchekoua replied:
Done !

mickaelandrieu commented on themes/_core/js/common.js:
```diff
@@ -28,3 +28,23 @@ export function psShowHide () {
+/**
+ * This function returns the value of the requested parameter from the URL
+ * @param {string} paramName - the name of the requested parameter
+ * @returns {string|null}
```
```diff
- @returns {string|null}
+ @returns {string|null|object}
```

mickaelandrieu reviewed (changes requested):
Some minor fixes, after changes good to me :+1:

vincentbz commented:
Hi, 

I still have a small issue if the address KO is not the default one. The error is not displayed, see gif:

![champ obligatoire adresse](https://user-images.githubusercontent.com/7567060/30599628-58a117ba-9d5d-11e7-90d5-d829f11ffff3.gif)


Thanks !

fatmaBouchekoua commented:
Hello @vincentbz ,
I've made some changes (That was a regression after the last modification)
Can you retest please ?

eternoendless commented on classes/Customer.php:
```diff
@@ -559,7 +559,8 @@ public static function resetAddressCache($idCustomer, $idAddress)
     {
-        $shareOrder = (bool) Context::getContext()->shop->getGroup()->share_order;
+        $group = Context::getContext()->shop->getGroup();
+        $shareOrder = isset($group->share_order) ? (bool) $group->share_order : false;
```
`false` should probably be replaced by `0`

eternoendless commented on classes/checkout/CartValidator.php:
```diff
@@ -0,0 +1,110 @@
+    /** @var Cart  */
+    private $cart;
+
+
```
Please avoid two empty lines in a row

eternoendless commented on controllers/front/OrderController.php:
```diff
@@ -150,7 +151,10 @@ protected function bootstrap()
+        $cartValidator = new CartValidator($this->context);
+        $customer = $this->context->customer;
+
+        $data['checksum'] = (is_array($cartValidator->hasValidAddresses()) && !$customer->isGuest()) ? null : $this->cartChecksum->generateChecksum($this->context->cart);
```
```php
is_array($cartValidator->hasValidAddresses())
```

Here you only need to know if all addresses are valid. Why does your method return an array with the first invalid address?

eternoendless replied:
It's probably better to test `!$customer->isGuest()` first, for better performance

eternoendless replied:
Why do you set `checksum` to `null`? Shouldn't we avoid writing data if the address is invalid?

LittleBigDev replied:
About checksum : previous persisted checksum needs to be invalidated anyway. That's why we set it to `null` when relevant.

eternoendless commented on classes/checkout/CartValidator.php:
```diff
@@ -0,0 +1,110 @@
+            try {
+                $address->getFields();
+            } catch (PrestaShopException $e) {
+                return array($idAddress);
```
This only returns the first invalid address, so why is it an array?

eternoendless commented on classes/checkout/CartValidator.php:
```diff
@@ -0,0 +1,110 @@
+
+        if (!($cart instanceof Cart)) {
+            return false;
+        }
```
If this method always needs a Cart, then why is not passed as a parameter? In that case, you wouldn't need the context anymore.

eternoendless commented on classes/checkout/CartValidator.php:
```diff
@@ -0,0 +1,110 @@
+        }
+
+        return $invalidAddresses;
+    }
```
This method checks the addresses of a customer. However, this class is called `CartValidator`. So why is this done here?

eternoendless commented on classes/checkout/CartValidator.php:
```diff
@@ -0,0 +1,110 @@
+        }
+
+        return true;
+    }
```
I think a better API would be the one that I suggested earlier:
* One method that tells you if all the address are valid (which is what you need to verify in the first place if I'm not mistaken).
* One method that tells you which address are invalid (which you need later in order to display the error).

Each method addresses a different concern.

eternoendless commented on controllers/front/OrderController.php:
```diff
@@ -168,7 +172,23 @@ protected function restorePersistedData(CheckoutProcess $process)
+        $customer = $this->context->customer;
+        $address = $cartValidator->hasValidAddresses();
+
+        if (is_array($address) && !$customer->isGuest()) {
```
this should be something like

```php
if (!$customer->isGuest() && $cartValidator->customerHasInvalidAddress($this->context->customer)) {
    $invalidAddresses = $cartValidator->getInvalidCustomerAddresses($this->context->customer));
    // ...
}

eternoendless commented on controllers/front/OrderController.php:
```diff
@@ -168,7 +172,23 @@ protected function restorePersistedData(CheckoutProcess $process)
+        $address = $cartValidator->hasValidAddresses();
+
+        if (is_array($address) && !$customer->isGuest()) {
+            $this->checkoutWarning['address'] = array(
```
What if both addresses are invalid?

eternoendless commented on themes/_core/js/checkout-address.js:
```diff
@@ -32,8 +36,74 @@ export default function () {
+  if ($('.js-address-error:visible').length > 0) {
+    let idFailureAddress = $(".js-address-error").prop('id').split('-').pop();
+    
+    $(".js-address-error:visible").each(function () {
```
Duplicate selector `$(".js-address-error:visible")`

eternoendless commented on classes/checkout/CheckoutAddressesStep.php:
```diff
@@ -202,18 +202,40 @@ public function handleRequest(array $requestParams = array())
+        $idAddressInvoice = (int) $this->getCheckoutSession()->getIdAddressInvoice();
+        $addressWarning = false;
+        $errorKey = 'address_error';
+        $checkoutWarning = isset($this->context->controller->checkoutWarning['address']) ? $this->context->controller->checkoutWarning['address'] : false;
```
I'm not a big fan of this, but I'm struggling to find another, cleaner way to allow the controller to pass this information along

vincentbz commented:
I've tested it and it's ok, thanks ! So, when the code review is completed, we'll do another test to validate that PR :)

xBorderie commented:
Once this one is merged, don't forget the Starter Theme port: https://github.com/PrestaShop/StarterTheme/pull/201

marionf commented:
Hello @fatmaBouchekoua 
I have a different invoice and delivery address.
The phone number is mandatory and neither of the 2 addresses has a phone number.
First, I try to edit the delivery address to add the phone number, I save.
Then, I am on the delivery step instead of the address step.
If I click on the address step, I see I have only one address selected for delivery & invoice.

I attached a screenrecord on the forge ticket.

LittleBigDev committed:
CartValidator is actually an address validator. So -> AddressValidator

LittleBigDev committed:
Addressed @eternoendless comments.
For the "biggest" comments, I had to refactor the validator class and the controller's behavior.

LittleBigDev committed:
Addressed @eternoendless comments for JS part

LittleBigDev committed:
Addressed @eternoendless + codacy comments

eternoendless commented:
Reviewed 5 of 15 files at r2, 1 of 3 files at r3, 4 of 4 files at r4, 3 of 3 files at r5.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[classes/Customer.php, line 564 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/7975#-Kw_UAgeGY2Nt5WWlq8I:-Kw_UAgeGY2Nt5WWlq8J:b-x6l2hx) ([raw file](https://github.com/prestashop/prestashop/blob/c1996e6d9c1de6d647720341f5968fd4a53001a1/classes/Customer.php#L564)):*
> ```PHP
>         $group = Context::getContext()->shop->getGroup();
>         $shareOrder = isset($group->share_order) ? (bool) $group->share_order : 0;
>         $cacheId = 'Customer::getAddresses'.(int) $this->id.'-'.(int) $idLang.'-'.$shareOrder;
> ```

Actually, after looking more closely, we should always cast `$shareOrder` to `int` so that we don't risk having an empty string.

---

*[classes/checkout/AddressValidator.php, line 28 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/7975#-Kw_USI0PGNphNwRlieQ:-Kw_USI0PGNphNwRlieR:btdk708) ([raw file](https://github.com/prestashop/prestashop/blob/c1996e6d9c1de6d647720341f5968fd4a53001a1/classes/checkout/AddressValidator.php#L28)):*
> ```PHP
> 
> /**
>  * Class AddressValidatorCore
> ```

Please describe the purpose of the class (otherwise this Phpdoc is useless)

---

*[classes/checkout/AddressValidator.php, line 75 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/7975#-Kw_VDNjeteHDHx0G8Wm:-Kw_VDNjeteHDHx0G8Wn:bdcvya6) ([raw file](https://github.com/prestashop/prestashop/blob/c1996e6d9c1de6d647720341f5968fd4a53001a1/classes/checkout/AddressValidator.php#L75)):*
> ```PHP
>      * The invalid address ids. Empty if everything is ok.
>      */
>     public function validateCustomerAddresses(Customer $customer, Context $context)
> ```

Why not simply pass the language id instead of `Context`?

---

*[classes/checkout/AddressValidator.php, line 88 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/7975#-Kw_VcESlMecKosvSk9m:-Kw_VcESlMecKosvSk9n:bi4rcvf) ([raw file](https://github.com/prestashop/prestashop/blob/c1996e6d9c1de6d647720341f5968fd4a53001a1/classes/checkout/AddressValidator.php#L88)):*
> ```PHP
>                     $invalidAddresses[] = (int)$address['id_address'];
>                 }
>             }
> ```

This foreach looks quite similar to the other one, can it be factored away?

---

*[controllers/front/OrderController.php, line 218 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/7975#-Kw_XpiW4VRqEiWIQ7d0:-Kw_XpiW4VRqEiWIQ7d1:b5i0ggh) ([raw file](https://github.com/prestashop/prestashop/blob/c1996e6d9c1de6d647720341f5968fd4a53001a1/controllers/front/OrderController.php#L218)):*
> ```PHP
>                 ),
>             );
>             $checksum                         = null;
> ```

I don't think this assignment is readable, it's too far away.

---

*[themes/_core/js/checkout-address.js, line 89 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/7975#-Kw_ZDXJckcsPEUGSe7H:-Kw_ZDXJckcsPEUGSe7I:b-v5effd) ([raw file](https://github.com/prestashop/prestashop/blob/c1996e6d9c1de6d647720341f5968fd4a53001a1/themes/_core/js/checkout-address.js#L89)):*
> ```JavaScript
>  * @param enabled
>  * @param id
>  * @param type
> ```

Missing types for these parameters

Example:
```js
/**
 * Assign the project to an employee.

 * @param {Object} employee - The employee who is responsible for the project.
 * @param {String} employee.name - The name of the employee.
 * @param {String} employee.department - The employee's department.
 *
 * @return {Project} Project instance
 */
Project.prototype.assign = function(employee) {
  // ...
}
```

You can find out more about JSDoc [here](http://usejsdoc.org/).

---

*[themes/_core/js/checkout-address.js, line 106 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/7975#-Kw_Zc_4kpHfTosBgiMR:-Kw_Zc_4kpHfTosBgiMS:blelpop) ([raw file](https://github.com/prestashop/prestashop/blob/c1996e6d9c1de6d647720341f5968fd4a53001a1/themes/_core/js/checkout-address.js#L106)):*
> ```JavaScript
>  */
> const switchConfirmAddresssesButtonState = function switchConfirmAddresssesButtonState() {
>   if ($(".js-address-error:visible").length > 0) {
> ```

This should be a parameter to the function

---

*[themes/classic/_dev/css/theme.scss, line 361 at r5](https://reviewable.io:443/reviews/prestashop/prestashop/7975#-Kw__CMTFjdTnVEx3VDK:-Kw__CMTFjdTnVEx3VDL:bsayarl) ([raw file](https://github.com/prestashop/prestashop/blob/c1996e6d9c1de6d647720341f5968fd4a53001a1/themes/classic/_dev/css/theme.scss#L361)):*
> ```SCSS
> .hide {
>   display: none;
> }
> ```

There should already be a `.hidden` class (see https://getbootstrap.com/docs/3.3/css/#helper-classes-show-hide), no need to add another one.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/7975)*
<!-- Sent from Reviewable.io -->

LittleBigDev committed:
build assets

eternoendless commented:
Reviewed 1 of 3 files at r3, 8 of 8 files at r6.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[install-dev/langs/en/data/configuration.xml, line 34 at r6](https://reviewable.io:443/reviews/prestashop/prestashop/7975#-Kwdjkp7pB8oiTcr9-EE:-Kwdjkp7pB8oiTcr9-EF:b1cfnxq) ([raw file](https://github.com/prestashop/prestashop/blob/20f0c04be406186427fa06386113fcae310d62ce/install-dev/langs/en/data/configuration.xml#L34)):*
> ```XML
>     <value>Out-of-Stock</value>
>   </configuration>
> </entity_configuration>
> ```

I don't think this change belongs to your PR :thinking: 

---

*[themes/_core/js/checkout-address.js, line 61 at r6](https://reviewable.io:443/reviews/prestashop/prestashop/7975#-KwdkAhby98jgPmMm8Ld:-KwdkAhby98jgPmMm8Le:bu1pl3f) ([raw file](https://github.com/prestashop/prestashop/blob/8d5ec6dfedc33c55f8c2f97ad42c42bf885bd80b/themes/_core/js/checkout-address.js#L61)):*
> ```JavaScript
>     }
> 
>     switchConfirmAddressesButtonState($(".js-address-error:visible").length > 0);
> ```

Watch out, you inverted the condition for enable/disable inside the function but you kept the original condition here.

---

*[themes/_core/js/checkout-address.js, line 82 at r6](https://reviewable.io:443/reviews/prestashop/prestashop/7975#-KwdjyX8PsoUU09IIIBU:-KwdjyX8PsoUU09IIIBV:bu1pl3f) ([raw file](https://github.com/prestashop/prestashop/blob/8d5ec6dfedc33c55f8c2f97ad42c42bf885bd80b/themes/_core/js/checkout-address.js#L82)):*
> ```JavaScript
>     });
>   }
>   switchConfirmAddressesButtonState($visibleAddressError.length > 0);
> ```

Watch out, you inverted the condition for enable/disable inside the function but you kept the original condition here.

---

*[themes/classic/templates/checkout/_partials/steps/addresses.tpl, line 72 at r6](https://reviewable.io:443/reviews/prestashop/prestashop/7975#-KwdkzueY_O9ajKVWMIe:-KwdkzueY_O9ajKVWMIf:b-ah5bgl) ([raw file](https://github.com/prestashop/prestashop/blob/8d5ec6dfedc33c55f8c2f97ad42c42bf885bd80b/themes/classic/templates/checkout/_partials/steps/addresses.tpl#L72)):*
> ```Smarty
>           <p class="alert alert-danger js-address-error" name="alert-delivery" id="id-failure-address-{$delivery_address_error.id_address}">{$delivery_address_error.exception}</p>
>         {else}
>           <p class="alert alert-danger js-address-error hide" name="alert-delivery">{l s="Your address is incomplete, please update it." d="Shop.Notifications.Error"}</p>
> ```

Class "hide" doesn't exist anymore, should be replace with "hidden"

---

*[themes/classic/templates/checkout/_partials/steps/addresses.tpl, line 116 at r6](https://reviewable.io:443/reviews/prestashop/prestashop/7975#-Kwdl4wxiXKLi_g_yx7e:-Kwdl4wxiXKLi_g_yx7f:b-k65cro) ([raw file](https://github.com/prestashop/prestashop/blob/8d5ec6dfedc33c55f8c2f97ad42c42bf885bd80b/themes/classic/templates/checkout/_partials/steps/addresses.tpl#L116)):*
> ```Smarty
>             <p class="alert alert-danger js-address-error" name="alert-invoice" id="id-failure-address-{$invoice_address_error.id_address}">{$invoice_address_error.exception}</p>
>           {else}
>             <p class="alert alert-danger js-address-error hide" name="alert-invoice" >{l s="Your address is incomplete, please update it." d="Shop.Notifications.Error"}</p>
> ```

Class "hide" doesn't exist anymore, should be replace with "hidden".
This may need to be fixed too on the Starter Theme

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/7975)*
<!-- Sent from Reviewable.io -->

LittleBigDev committed:
Addressed new @eternoendless comments

LittleBigDev committed:
Build assets

LittleBigDev commented:
@eternoendless rebase fixes are done and your comments are addressed. Ok for review.

eternoendless commented:
I think your last change introduced a regression

---

Reviewed 4 of 5 files at r8.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[themes/_core/js/checkout-address.js, line 47 at r8](https://reviewable.io:443/reviews/prestashop/prestashop/7975#-KwetvkEfdwuyJocbhNw:-KwetvkEfdwuyJocbhNx:bx9ps4) ([raw file](https://github.com/prestashop/prestashop/blob/12bfac8027da71ed11835ffcaa061b356e37b5fe/themes/_core/js/checkout-address.js#L47)):*
> ```JavaScript
>     let addressType = this.name.split('_').pop();
>     let $addressError = $('.js-address-error[name=alert-' + addressType + ']');
>     let $visibleAddressError = $('.js-address-error:visible');
> ```

I think this selector needs to be performed just before `switchConfirmAddressesButtonState()` below, because some elements are being hidden after this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/7975#-:-KweugxNZ9BtZZLEiEfY:b-id510u)*
<!-- Sent from Reviewable.io -->

LittleBigDev commented:
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[themes/_core/js/checkout-address.js, line 47 at r8](https://reviewable.io:443/reviews/prestashop/prestashop/7975#-KwetvkEfdwuyJocbhNw:-KwiEmhUXJGS3N5YDmMb:b-p0ghp6) ([raw file](https://github.com/prestashop/prestashop/blob/12bfac8027da71ed11835ffcaa061b356e37b5fe/themes/_core/js/checkout-address.js#L47)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

I think this selector needs to be performed just before `switchConfirmAddressesButtonState()` below, because some elements are being hidden after this.
</blockquote></details>

ok will fix

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/7975)*
<!-- Sent from Reviewable.io -->

LittleBigDev committed:
Fixed regression on checkout page (address error message visibility)

LittleBigDev commented:
Review status: 9 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

*[themes/_core/js/checkout-address.js, line 47 at r8](https://reviewable.io:443/reviews/prestashop/prestashop/7975#-KwetvkEfdwuyJocbhNw:-KwiFlP7HOAfo048le8B:b-5ghkrq) ([raw file](https://github.com/prestashop/prestashop/blob/12bfac8027da71ed11835ffcaa061b356e37b5fe/themes/_core/js/checkout-address.js#L47)):*
<details><summary><i>Previously, LittleBigDev (Bastien Bieri) wrote…</i></summary><blockquote>

ok will fix
</blockquote></details>

pushed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/7975)*
<!-- Sent from Reviewable.io -->

LittleBigDev committed:
Build assets

eternoendless commented:
Reviewed 1 of 1 files at r9, 2 of 2 files at r10.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/7975)*
<!-- Sent from Reviewable.io -->

LittleBigDev committed:
Fixed checkout address error message visibility
- `hidden` was not the correct class to use (correct one is `invisible`)
- anyway, jquery `.hide()` was not working if `invisible` class was set

eternoendless commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r11.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/7975#-:-Kwis3oMX4ns5iUBsj-1:bnfp4nl)*
<!-- Sent from Reviewable.io -->

eternoendless commented:
Thank you @fatmaBouchekoua @LittleBigDev

marionf commented:
Hello @eternoendless & @LittleBigDev 
We still have an issue here:

![capture du 2017-10-20 09-35-21](https://user-images.githubusercontent.com/13449658/31810313-9d27ac88-b57b-11e7-9a34-a25876db40ae.png)

For more details, please see the screenrecord attached on the forge ticket

LuigiMdg commented:
I've the same problem..
All fields are writed but he take to me this error and users can't go to complete the checkout..!

Luxurdk commented:
...HERE IS HOW TO FIX IT.... in ps 1.7.5 - getting rid off error 500 in checkout

Go to your file AddressFormat.php into the path Project_Folder/classes/AddressFormat.php, the list of requireFormFieldsList.
![46488500-20781d80-c803-11e8-80db-2222f9fd8f2b](https://user-images.githubusercontent.com/50981231/59085991-b0291b00-8900-11e9-85cb-cc73bfe96408.png)

And type the fields you want here.
For example 'phone',

If that is the field you want as required

Hope this helps.
